Add GHC 9.2 support

brandon-leapyear opened the pull request:
Related: #1475

akhesaCaro reviewed (changes requested):
Thanks for your contribution,

Here some additional tasks that mustn't be forgotten :
- [ ] Be sure that the Nix can handle GHC 9.2 
- [ ] Add the parameter in the nix/README
- [ ] Add a check with Cabal on the CI (Only the stack check have been updated on the CI) - see https://github.com/haskell-servant/servant/blob/master/.github/workflows/master.yml

Are you okay with doing that?

Note : The PR (#1475) must be merged and this PR rebased before integrating it to the code base.

brandon-leapyear commented:
I rebased and updated nix/README + github actions. I don't use Nix, so I won't be able to help with the Nix point, and I don't know what you mean by "Cookbook"

brandon-leapyear commented on servant/src/Servant/Links.hs:
```diff
@@ -649,7 +649,7 @@ simpleToLink _ toA _ = toLink toA (Proxy :: Proxy sub)
 instance TypeError (PartialApplication HasLink arr) => HasLink ((arr :: a -> b) :> sub)
   where
-    type MkLink (arr :> sub) _ = TypeError (PartialApplication HasLink arr)
+    type MkLink (arr :> sub) _ = TypeError (PartialApplication (HasLink :: * -> Constraint) arr)
```
This was added 2 weeks ago: https://github.com/haskell-servant/servant/commit/42b7d0eb9bc36e02fc4aa2656bcbde9aa540aaee

GHC 9.2 is stricter about this:

> GHC is stricter about checking for out-of-scope type variables on the right-hand sides of associated type family instances that are not bound on the left-hand side.

[source](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/9.2.1-notes.html)

josephcsible replied:
aab7e0d added another instance of this problem: https://github.com/haskell-servant/servant/blob/7a770b5a1e0fc250c5f19c0629329e28b69151c0/servant-server/src/Servant/Server/Internal.hs#L824
`HasServer` needs to become `(HasServer :: * -> [*] -> Constraint)` (and `Constraint` needs to be added to the import list).

brandon-leapyear commented:
@ysangkok I changed GHC 9.2 to run with Cabal 3.6, but I'm still getting errors. I don't use Cabal myself, can you help fix that?

akhesaCaro commented:
Okay I will try on my side and keep you posted.

gdeest commented:
I am not entirely making sense of the Cabal failure ; it seems to me that Cabal commits to a bound-to-fail plan because the constraints it tries to satisfy do not match pre-installed versions, but I am not entirely sure.

I am slightly concerned with the amount of `allow-newer` stanzas we have in the `cabal.project` files:

```
-- https://github.com/chordify/haskell-servant-pagination/pull/12
allow-newer: servant-pagination-2.2.2:servant
allow-newer: servant-pagination-2.2.2:servant-server

allow-newer: servant-js:servant

-- ghc 9
allow-newer: tdigest:base

-- ghc 9.2
allow-newer: basement:base
allow-newer: binary:bytestring
allow-newer: generics-sop:base
allow-newer: generics-sop:ghc-prim
allow-newer: generics-sop:template-haskell
allow-newer: http-media:base
allow-newer: lucid:base
allow-newer: memory:base
allow-newer: openssl-streams:bytestring
allow-newer: servant-js:base
allow-newer: servant-js:base-compat
allow-newer: servant-js:lens
allow-newer: servant-multipart:bytestring
allow-newer: servant-swagger:base-compat
allow-newer: singleton-bool:base
allow-newer: sop-core:base
allow-newer: swagger:base
allow-newer: swagger2:base
allow-newer: swagger2:base-compat-batteries
allow-newer: swagger2:bytestring
allow-newer: swagger2:lens
allow-newer: swagger2:template-haskell
allow-newer: swagger2:time
```

I am guessing we don't really have a choice because these libraries do not have 9.2-compatible releases yet >

gdeest commented:
I have managed to get past Cabal configure, but `basement` fails to build with GHC 9.2. We'll have to exclude some things from the 9.2 build.

tchoutri commented:
Yes, the Hackage & Stackage folks have had some issues with [basement](https://github.com/haskell-infra/hackage-trustees/issues/330) and [foundation](https://github.com/commercialhaskell/stackage/issues/6400#issuecomment-1019001230), looks like the 9.2 support for these libraries will come one day™

brandon-leapyear commented:
@gdeest in the future, can you let me know before you push changes to my fork? Thanks!

josephcsible commented on servant-server/src/Servant/Server/Internal.hs:
```diff
@@ -821,7 +823,7 @@ instance (HasContextEntry context (NamedContext name subContext), HasServer subA
 instance TypeError (PartialApplication HasServer arr) => HasServer ((arr :: a -> b) :> sub) context
   where
-    type ServerT (arr :> sub) _ = TypeError (PartialApplication HasServer arr)
+    type ServerT (arr :> sub) _ = TypeError (PartialApplication (HasServer :: Type -> Constraint) arr)
```
Shouldn't that be `Type -> [Type] -> Constraint`?

ysangkok commented:
I think it is a bad idea to make the default `stack.yaml` on the master branch refer to unreleased versions of libraries (e.g. `lens`, `foundation`, etc). I see how it makes sense to do this for development purposes, though. But if this branch isn't supposed to be merged, maybe it would be clearer to mark it "Draft"? After dependencies are fixed or removed, it could be marked ready for review again.

ysangkok commented on stack.yaml:
```diff
@@ -13,8 +15,34 @@ packages:
+  subdirs:
+    - warp
+
+# https://github.com/ekmett/lens/issues/990
+- github: ekmett/lens
+  commit: db49dcf9c7d757baa6a1f42111edd8e2729a9c9c
```
This can be removed since a fixed lens version was released.

ysangkok commented on stack.yaml:
```diff
@@ -13,8 +15,34 @@ packages:
+- attoparsec-0.14.2
+- base-compat-0.12.1
+- base-compat-batteries-0.12.1
+- github: yesodweb/wai
+  commit: 02441350b50558f628158ab50b023b5328f38a41
+  subdirs:
+    - warp
```
This should be unnecessary since https://github.com/yesodweb/wai/commit/493610faeb7c2dedd28ecb43e6a6adfb6e0f4eac was released.

brandon-leapyear commented:
I intended for the PR to be merged now. I wasn't sure how Stack is used in this project, e.g. if the standard build workflow is via Cabal, and the stack config is just there for reference. I've moved the Stack stuff into a new `stack-ghc-9.2.yaml` config.

I would prefer to not have a PR that I maintain open for an indefinite period of time. If you'd prefer to not merge any aspect of this PR at this time, feel free to close.

ysangkok commented on stack-ghc-9.2.yaml:
```diff
@@ -0,0 +1,41 @@
+- github: haskell-crypto/cryptonite
+  commit: 3b081e3ad027b0550fc87f171dffecbb20dedafe
+
+# https://github.com/aesiniath/http-streams/pull/126
+- github: aesiniath/http-streams
+  commit: 86cd7f970d747ee653328e59176d451780641992
```
These lines are now unnecessary since http-streams 0.8.9.6 have been released containing aesiniath/http-streams/pull/126 (which you submitted)

ysangkok commented on stack-ghc-9.2.yaml:
```diff
@@ -0,0 +1,41 @@
+- servant-swagger/
+
+extra-deps:
+# https://github.com/GetShopTV/swagger2/pull/231
+- github: GetShopTV/swagger2
+  commit: 1313547f8ebe594198a1486f0592a962da83e8c6
```
These lines are now unnecessary since swagger2 version 2.8 has been released.

ysangkok commented on cabal.project:
```diff
@@ -77,3 +69,35 @@ allow-newer: servant-multipart-api:servant
+allow-newer: singleton-bool:base
+allow-newer: sop-core:base
+allow-newer: swagger:base
+allow-newer: swagger2:base
+allow-newer: swagger2:base-compat-batteries
+allow-newer: swagger2:bytestring
+allow-newer: swagger2:lens
+allow-newer: swagger2:template-haskell
+allow-newer: swagger2:time
```
These lines are now unnecessary because swagger2 version 2.8 has been released with adjusted bounds.

ysangkok commented on cabal.project:
```diff
@@ -77,3 +69,35 @@ allow-newer: servant-multipart-api:servant
+allow-newer: servant-swagger:base-compat
+allow-newer: servant-swagger:base-compat
+allow-newer: servant-swagger:Cabal
+allow-newer: singleton-bool:base
```
This line is unnecessary since a revision was published: https://hackage.haskell.org/package/singleton-bool-0.1.6/revisions/

ysangkok commented on cabal.project:
```diff
@@ -77,3 +69,35 @@ allow-newer: servant-multipart-api:servant
+allow-newer: servant-swagger:base-compat
+allow-newer: servant-swagger:Cabal
+allow-newer: singleton-bool:base
+allow-newer: sop-core:base
```
This line is unneceessary since sop-core 0.5.0.2 was released: https://hackage.haskell.org/package/sop-core

ysangkok commented on cabal.project:
```diff
@@ -77,3 +69,35 @@ allow-newer: servant-multipart-api:servant
+-- ghc 9.2
+allow-newer: basement:base
+allow-newer: binary:bytestring
+allow-newer: generics-sop:base
+allow-newer: generics-sop:ghc-prim
+allow-newer: generics-sop:template-haskell
```
These lines are unnecessary since generics-sop 0.5.1.2 has been released: https://hackage.haskell.org/package/generics-sop

ysangkok commented on cabal.project:
```diff
@@ -77,3 +69,35 @@ allow-newer: servant-multipart-api:servant
+allow-newer: servant-swagger:Cabal
+allow-newer: singleton-bool:base
+allow-newer: sop-core:base
+allow-newer: swagger:base
```
Why was this added? The bounds are `>=4.6 && <5.0` which is wide enough.

brandon-leapyear replied:
🤷 it was needed when I opened this PR 3 months ago

ysangkok commented on cabal.project:
```diff
@@ -77,3 +69,23 @@ allow-newer: servant-multipart-api:servant
+
+-- ghc 9.2
+allow-newer: basement:base
+allow-newer: binary:bytestring
+allow-newer: http-media:base
+allow-newer: lucid:base
```
These have been updated and can be removed: binary, http-media, lucid

ysangkok commented on cabal.project:
```diff
@@ -77,3 +69,23 @@ allow-newer: servant-multipart-api:servant
+allow-newer: servant-multipart:http-media
+allow-newer: servant-multipart:lens
+allow-newer: servant-multipart:resourcet
+allow-newer: servant-swagger:base
+allow-newer: servant-swagger:base-compat
+allow-newer: servant-swagger:base-compat
+allow-newer: servant-swagger:Cabal
```
servant-swagger has been updated and these lines can be removed

shlevy commented on servant-server/src/Servant/Server/Internal.hs:
```diff
@@ -821,7 +823,7 @@ instance (HasContextEntry context (NamedContext name subContext), HasServer subA
 instance TypeError (PartialApplication HasServer arr) => HasServer ((arr :: a -> b) :> sub) context
   where
-    type ServerT (arr :> sub) _ = TypeError (PartialApplication HasServer arr)
+    type ServerT (arr :> sub) _ = TypeError (PartialApplication (HasServer :: Type -> [Type] -> Constraint) arr)
```
Can we get a quick patch release with just this change included? With this I think GHC 9.2.1 compilation will succeed if allow-newer is set and fixed versions of cryptonite, memory, and basement are included in the cabal.project.

ysangkok replied:
It was already merged as part of

* haskell-servant/servant/pull/1525

ysangkok commented on cabal.project:
```diff
@@ -77,3 +69,16 @@ allow-newer: servant-multipart-api:servant
+allow-newer: basement:base
+allow-newer: memory:base
+allow-newer: openssl-streams:bytestring
+allow-newer: servant-js:base
+allow-newer: servant-js:base-compat
+allow-newer: servant-js:lens
```
These lines can be removed since [servant-js was revised](https://hackage.haskell.org/package/servant-js-0.9.4.2/revisions/).

ysangkok commented on cabal.project:
```diff
@@ -77,3 +69,13 @@ allow-newer: servant-multipart-api:servant
+allow-newer: servant-multipart:bytestring
+allow-newer: servant-multipart:bytestring
+allow-newer: servant-multipart:http-media
+allow-newer: servant-multipart:lens
```
* https://github.com/haskell-servant/servant-multipart/pull/59

ysangkok replied:
@brandon-leapyear These ones for servant-multipart and the one for openssl-streams can be removed now. If you rebase, the diff will also get smaller now because allow-newers were removed on master.

ysangkok commented on cabal.project:
```diff
@@ -77,3 +69,13 @@ allow-newer: servant-multipart-api:servant
+-- ghc 9.2
+allow-newer: basement:base
+allow-newer: memory:base
+allow-newer: openssl-streams:bytestring
```
* snapframework/openssl-streams/pull/6

brandon-leapyear commented:
@ysangkok can we get this merged anytime soon? I'd prefer to not be maintaining this long-term

ysangkok commented:
@brandon-leapyear I am not a project maintainer, so it is not my decision to make. But I have submitted #1555 which achieves GHC 9.2 compatibility for a subset of packages that work on GHC 9.2, and marks the others as unbuildable. I prefer this approach due to how it doesn't legitimize unofficial forks of dependencies (as [previously mentioned](https://github.com/haskell-servant/servant/pull/1482#issuecomment-1028160110)).

brandon-leapyear commented:
Fine by me. Can you link the same issue in your PR?

On Wed, Mar 9, 2022, 11:06 AM Janus Troelsen ***@***.***>
wrote:

> @brandon-leapyear <https://github.com/brandon-leapyear> I am not a
> project maintainer, so it is not my decision to make. But I have submitted
> #1555 <https://github.com/haskell-servant/servant/pull/1555> which
> achieves GHC 9.2 compatibility for a subset of packages that work on GHC
> 9.2, and marks the others as unbuildable. I prefer this approach due to how
> it doesn't legitimize unofficial forks of dependencies (as previously
> mentioned
> <https://github.com/haskell-servant/servant/pull/1482#issuecomment-1028160110>
> ).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell-servant/servant/pull/1482#issuecomment-1063261363>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGUC75NJBNCEESHLHBBDVUTU7DZENANCNFSM5ICWIIJQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

maksbotan commented:
@ysangkok @brandon-leapyear please check if this PR is still relevant after merge of #1555.

And thanks for your work!
